scripts/build_images/build.py

```diff
@@ -0,0 +1,205 @@
+}
+
+BASE_IMAGES = list(set(BASE_IMAGE_MAPPING.values()))
+
+images = get_env_or_exit("IMAGES")
+tag = get_env_or_exit("TAG")
+metadata_dir = os.environ.get("METADATA_DIR")
+should_push = os.environ.get("PUSH_IMAGES", "false").lower() == "true"
+
+if not metadata_dir:
+    print("No manifest directory specified; manifests will not be created.")
+
+
+def get_image_path(image: str) -> str:
+    if not image.startswith("prairielearn/"):
+        raise ValueError(f"Cannot build non-PrairieLearn image: {image}")
+
+    image = image[len("prairielearn/") :]
+
+    if image.startswith("workspace-"):
+        image = image[len("workspace-") :]
+        return f"workspaces/{image}"
+
+    if image.startswith("grader-"):
+        image = image[len("grader-") :]
+        return f"graders/{image}"
+
+    if image == "plbase":
+        return "images/plbase"
+
+    if image == "executor":
+        return "images/executor"
+
+    if image == "prairielearn":
+        return "."
+
+    raise ValueError(f"Cannot build unknown image: {image}")
+
+
+def get_current_platform() -> str:
+    result = subprocess.run(
+        ["docker", "version", "--format", "json"],
+        capture_output=True,
+        check=True,
+    )
+    version_data = json.loads(result.stdout)
+    return f"{version_data['Server']['Os']}/{version_data['Server']['Arch']}"
+
+
+def check_path_modified(path: str) -> bool:
+    branch = (
+        subprocess.run(
+            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
+            capture_output=True,
+            check=True,
+        )
+        .stdout.decode()
+        .strip()
+    )
+
+    # If this script is being run *on* the master branch, then we want to diff
+    # with the previous commit on master. Otherwise, we diff with master itself.
+    diff_branch = "HEAD^1" if branch == "master" else "remotes/origin/master"
+
+    diff_result = subprocess.run(
+        ["git", "diff", "--exit-code", f"{diff_branch}..HEAD", "--", path],
+        capture_output=True,
+        check=False,
+    )
+
+    return diff_result.returncode != 0
+
+
+# Note that base images must come first in the list so that they're built first.
+image_list = images.split(",")
+has_base_image = any(image in BASE_IMAGES for image in image_list)
+
+
+platform = get_current_platform()
+
+built_images: set[str] = set()
+
+with local_registry(REGISTRY_NAME) if has_base_image else nullcontext():
+    for image in image_list:
+        is_base_image = image in BASE_IMAGES
+        base_image = BASE_IMAGE_MAPPING.get(image)
+        base_image_built = base_image in built_images
+        image_path = get_image_path(image)
+        was_modified = check_path_modified(image_path)
+
+        if not was_modified and not base_image_built:
+            print(f"Skipping {image} because it hasn't changed.")
+            continue
+
+        built_images.add(image)
+
+        # Make temporary files for the metadata.
+        with tempfile.NamedTemporaryFile(delete=False) as metadata_file:
```

Comment:
Shouldn't use a context manager here, just `metadata_file = ...`. Or, put the code inside the with block.

Reply:
The ruff warning wants you to put your code (the `...`) in the with block. https://docs.astral.sh/ruff/rules/open-file-with-context-handler/